### PR TITLE
[FIX] account: always post message on using Send and Print

### PR DESCRIPTION
**Pull request by yelizariev**
Invoices can be sent via Send and Print action. The action works differently depending on number of invoices selected: a single invoice is sent as a regular message, while multiple invoices are posted as internal note and are sent immediatly [1][2]. There are two related issues on such implementation:

1. Mails might be sent twice because of concurency error. See #92740

2. Internal note has no envelope icon and hence user cannot verify that the mail is delivered. See linked tickets

Both issues are not good for such important operation as invoicing the customers.

Fix it by processing every invoice individually. This should also improve performance of the wizard, because mails are not sent immediatly.

[1]: https://github.com/odoo/odoo/blob/0aff8bb9484a23c0d875d3b12259dd4e0d51e716/addons/account/wizard/account_invoice_send.py#L36 [2]: https://github.com/odoo/odoo/blob/0aff8bb9484a23c0d875d3b12259dd4e0d51e716/addons/mail/wizard/mail_compose_message.py#L266-L267

opw-2745412
opw-3087711

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Review by rousseldenis — APPROVED** [review]
LGTM. Code and functional review

**Review by baimont — APPROVED** [review]

**Comment by epictete** [comment]
From an accounting point of view it looks good to me, but I'm far from familiar with the implications it might have in the mass mailing process (which it seems to be bypassing), so I think it would be nice if someone from the RD-Marketing team could have a look :wink:

**Comment by tde-banana-odoo** [comment]
Looks like a terrible idea to me.

**Comment by yelizariev** [comment]
> Looks like a terrible idea to me.

Please elaborate

**Comment by tde-banana-odoo** [comment]
Hello,

Yep, sorry, had some stuff to do :D in short: you are modifying the behavior in stable and performance will be much worse. Posting != Mailing . Those are two different concepts.

Posting: post on a document, notify followers, send emails or inbox notifications, track status of notifications, ... This is quite costly as a lot of side records are involved. The layout you expect actually comes from the post on invoice document that sends email notifications with a layout.

Mailing: create emails in batch, with recipients coming from template definition. In that case, we expect the template to be complete and less over computation is performed. It is also more optimized for batch.

Note that current account.invoice.send wizard code is a bit weird, as it is a mix of composition (inherits) and manual management (hardcoding some values, calling onchange, creating them on the fly). I expect this code to be simplified in master after https://github.com/odoo/odoo/pull/107356 as while modifying the mail composer, having a look at the account composer is not that much more work.

Concerning code: you are copying the composer in a loop... if you send an email to 1000 invoices, you will create 1000 composer in a loop. This seems not efficient to me.

In short: this is a functional change, it has a lot of impact (performance, results), and moreover your change is very costly form performances point of view.

Cheers,

**Comment by thiryr** [comment]
Fyi, for point 2, we have a PR to add notifications on mass mails so I think that issue should be solved when we merge it.

**Comment by yelizariev** [comment]
> Fyi, for point 2, we have a PR to add notifications on mass mails so I think that issue should be solved when we merge it.

is it for master?

**Comment by yelizariev** [comment]
> Concerning code: you are copying the composer in a loop... if you send an email to 1000 invoices, you will create 1000 composer in a loop. This seems not efficient to me.

It's not efficient, but we are limited on what we can do in stable. You are right, that it may increase memory usage, but who cares about temporary memory peak if it works and allows user to send and properly track 1000 invoices. 

It is a functional change indeed, but it's the only workaround I see for the mentioned issues. We have to fix those issues or find a workaround at any cost , because sending invoices is just too important operation for any business. If company sends 1000 invoices and loose just 1% of emails, then their ERP sucks.

**Comment by tde-banana-odoo** [comment]
Hello @yelizariev ,

This is a current limitation of the mail framework that applies to all models since Odoo exists. This does not qualify for a bug in my opinion, this is a technical limitation. Your fix has huge drawbacks that impacts notably a lot performance. You wanted my opinion, you have it. I let you do whatever you want with accounting specific use case. I won't r+ or r- anything related to accounting in any way.

For what is relevant to mail itself, its behavior is being improved in master, as we don't think it is safe to change that code in stable.

Cheers,

**Comment by rousseldenis** [comment]
> Hello @yelizariev ,
> 
> This is a current limitation of the mail framework that applies to all models since Odoo exists. This does not qualify for a bug in my opinion, this is a technical limitation. Your fix has huge drawbacks that impacts notably a lot performance. You wanted my opinion, you have it. I let you do whatever you want with accounting specific use case. I won't r+ or r- anything related to accounting in any way.
> 
> For what is relevant to mail itself, its behavior is being improved in master, as we don't think it is safe to change that code in stable.
> 
> Cheers,

 @tde-banana-odoo The drawback is you force users to validate one by one invoices if they want a feedback on the sent mail.

This is deployed on production without problem.

**Comment by C3POdoo** [comment]
Dear @yelizariev,

Thank you for your contribution but the version 14.0 is no longer supported.
We only support the last 3 stable versions so no longer accepts patches into this branch.

We apology if we could not look at your request in time.
If the contribution still makes sense for the upper version, please let us know and do not hesitate to recreate one for the recent versions. We will try to check it as soon as possible.

_This is an automated message._
